### ISSUE TITLE
Do not auto idle if user input is active

### ISSUE THREAD

**Issue by UlyssesZh** [Vencord/Vesktop#396]
**Describe the bug**

Discord web goes to auto idle when the web page is unfocused for some time. Discord desktop client goes to auto idle when the user input is inactive for some time regardless of whether the client window is focused. The latter is much more sensible than the former. While I see there are limitations for a web page, I think this should be considered a bug because Vesktop is not behaving like what one would expect a desktop client to do.

**To Reproduce**

**Expected behavior**

**Screenshots**

**Desktop (please complete the following information):**
 - OS/Distro: NixOS 23.11
 - Desktop Environment (linux only): GNOME 45
 - Version: 1.5.0

**Command line output**

**Additional context**

**Comment by NickelWW** [comment]
I noticed this behavior as well, and I agree

Desktop:
- OS: Windows 11
- Version: 1.5.0

**Comment by NickelWW** [comment]
> I'm experiencing this issue as well. It's likely a limitation of Electron and it likely wouldn't be worth it to try and fix. I think a more sensible solution would be to have the option to disable idle functionality all together as suggested in #451.

I don't think so, because stock discord desktop is also an electron app and it doesn't have this issue...

**Comment by nan60** [comment]
I may have phrased that wrong. Discord is Electron + a bunch of proprietary stuff, such as their rich status API and screen sharing (which is why it didn't work with Wayland and why it doesn't have sound in the official client). I would assume (I am assuming here) that this would be another custom component that would instead have to be implemented manually.

**Comment by UlyssesZh** [comment]
It seems that Vendicated/Vencord#2342 can partially solve this problem.

**Comment by Vendicated** [comment]
so just for clarification since I myself got confused, this issue is about vesktop still idling even if the user is actively using their mouse/keyboard/...

vesktop could definitely fix that and copy the discord desktop behaviour of only idling if there's no input. but likely only on X11/XWayland, not Wayland

**Comment by UlyssesZh** [comment]
Yes, this is what I meant.

It's good to hear that this is possible.

**Comment by Vendicated** [comment]
Discord Desktop logic (cleaned up & simplified, search for `"IdleStore` to find this in code):
```js
function isSystemIdle() {
  return suspended || locked || (isAndroid() && idkIrrelevantWhoCares);
}

function checkIdleNative() {
	let checkIdleTime = t => {
	    idleSince = Math.max(Date.now() - t, idleSince),
	    checkIdleAFK(),
	    setTimeout(checkIdleNative, 10 * d.Z.Millis.SECOND)
	}
    DiscordNative.remotePowerMonitor.getSystemIdleTimeMs()
		.then(checkIdleTime);
}

checkIdleNative();

function handleEvent() {
	// [...]
	checkIdleOrAfk();
}

DiscordNative.remotePowerMonitor.on("resume", ()=>{
    suspended = false,
    handleEvent()
}
),
DiscordNative.remotePowerMonitor.on("suspend", ()=>{
    suspended = true,
    handleEvent()
    disconnectFromVoiceChannel();
}
),
DiscordNative.remotePowerMonitor.on("lock-screen", ()=>{
    locked = true,
    handleEvent()
}
),
DiscordNative.remotePowerMonitor.on("unlock-screen", ()=>{
    locked = false,
    handleEvent()
}
```

we will have to port this logic. we will have to patch the isSystemIdle function to replace `locked` and `suspended` with our own variables. we can call `handleEvent()` by dispatching a `OVERLAY_SET_NOT_IDLE` flux event

**Comment by llyyr** [comment]
> vesktop could definitely fix that and copy the discord desktop behaviour of only idling if there's no input. but likely only on X11/XWayland, not Wayland

This can be done on wayland too using the ext-idle-notify protocol. We don't care about checking for mouse/keyboard activity, we just want to know if the user is idle or not and the compositor can provide this info

**Comment by Vendicated** [comment]
The reason it's not working is because it's simply not implemented in Vesktop. It would have to be ported over from Discord just like I described above

**Comment by llyyr** [comment]
> The reason it's not working is because it's simply not implemented in Vesktop. It would have to be ported over from Discord just like I described above

For non-Wayland platforms sure, the issue still exists that Electron can't detect user idle on Wayland. I found the issue tracking this on Electron's side https://github.com/electron/electron/issues/27912

I can't see how you intend to actually poll how long the user has been idle for.

**Comment by Jamesking56** [comment]
Bumping this as its still happening

**Comment by Novack** [comment]
Found this issue looking into why the status was inconsistent compared to the official client. Im trying a plugin for now, to see if it can help. Im on X11.

<img width="884" height="291" alt="Image" src="https://github.com/user-attachments/assets/c7798bc7-19fa-4ed4-a16b-e07f6530a9be" />
